### ISSUE TITLE
ENH: Support float32 for KDTree

### ISSUE THREAD

**Issue by dkbarn** [scipy/scipy#18467]
### Is your feature request related to a problem? Please describe.

Currently the implementation of `scipy.spatial.KDTree` only allows input data of type `np.float64`. If your data is anything other than this, it is copied into an `np.float64` array. KDTree should support `np.float32` as input as well, which would allow for very large datasets at half the memory consumption and with negligible impact on accuracy.

### Describe the solution you'd like.

Both the data and query input for `scipy.spatial.KDTree` can be `np.float32` with `copy=False`. In these cases, the KDTree would use 32-bit floating point arithmetic internally, and return distance values that are also `np.float32`.

### Describe alternatives you've considered.

I've tried using `sklearn.neighbors.KDTree` as well as `pykdtree.kdtree.KDTree`. The former does not have as many features as scipy's implementation, lacking the variety of query methods, and while the latter does support `np.float32`, it is limited to a maximum vector size of 127.

**Comment by tylerjereddy** [comment]
Maybe cc @sturlamolden for feedback/if this has been talked about before.

**Comment by sturlamolden** [comment]
A kd-tree suffers from the **curse of dimensionality** and is only effective for low dimensional data. The amount of work it needs to do grows exponentially with the number of dimensions. 

See e.g. Section 2.1 here:

https://www.cs.yale.edu/homes/el327/datamining2013aFiles/11_nearest_neighbor_search.pdf

If you have random points in ```[0,1]**d``` a kd-tree is expected to visit ```2**d``` partitions to find a neighbor. 

If you have a 127-dimensional space the data is pathological for a kd-tree as it is expected to visit an astronomic number of partitions. You will be better off with brute-force linear search. Any limit on the maximum vector length is not a reason to prefer ```scipy.spatial.KDTree``` over ```sklearn.neighbors.KDTree```.

**Comment by dkbarn** [comment]
I understand. However this issue is not really about high dimensionality, it's about memory efficiency. Implementing a KDTree in float64 only is prohibitively restrictive for large datasets. And those datasets can be large in the number of rows alone, with dimension < 127.

As I stated, sklearn's KDTree is not a sufficient alternative because it doesn't provide the same range of query methods that scipy's provides, such as query_ball_point and query_ball_tree.

Regardless, it looks like sklearn's KDTree is ALSO only float64, as this open issue indicates: https://github.com/scikit-learn/scikit-learn/issues/15474

**Comment by sturlamolden** [comment]
Scikit-learn has a kd-tree written entirely in Cython. SciPy uses a backend in C++. This allows SciPy’s kd-tree to release the GIL, which means it can be used for things like interactive graphics and computer games. That is, it does not lock up the interpreter, and cause a GUI to freeze. This is the main difference between them. If you are doing modern hype like data science, machine learning, AI you will probably not observe any difference – but if you are coding a flight simulator you will. At least this is how it used to be, I have not looked at the sklearn code for a while.

As for supporting float32 it is a major rewrite. However, all C++ code can be templated and you can use Cython as a bloatware generator. Support for float32 would be nice, but I do not see the point if the reason is to support 127-dimensional spaces. If it is support computer graphics, which in general is done in single precision, there is a good case for it. At the same time one could also consider to support integer data. 

But it will be a major undertaking - totally doable though.

**Comment by sturlamolden** [comment]
I believe their range query is called query_range. The name query_ball_point for a range query is a misnomer. 

There is actually ony two or three queries a kd-tree can do: Neighbor query, range query, and pairs query (though the latter is a special case of neighbor query).

**Comment by sturlamolden** [comment]
Actually their range query is called query_radius.

**Comment by dkbarn** [comment]
I just wanted to point out that your suggestion of using a brute force approach for high dimensionality is equally problematic in terms of memory consumption.

This is because the `scipy.spatial.distance.cdist` function also only works with float64 data and always returns float64. So, supposing you have a very large (in rows) dataset, which you are storing at float32 or float16 (in order to sit comfortably in memory, and because your use case is tolerant of single- or half-precision accuracy), the result of running a brute force cdist on this data will produce an enormous float64 distance matrix.

In short, we need a memory efficient way of computing euclidean distances, either through kd-trees or via brute force, for very large datasets.

**Comment by dkbarn** [comment]
After a cursory glance through the open issues in this repo, I can see that I'm not alone in wishing for more memory efficient computation from this library.

#11280 asks for control over dtype for `pdist`
#15065 is another request for single-precision float in kd-trees
#15602 asks for dtype control in all `optimize/_differentiable_functions.py` functions
#18090 asks for dtype control in Dijkstra's distance matrix

All these issues cite the same reason: "forcing float64 conversion becomes prohibitive when dealing with large graphs"

**Comment by sturlamolden** [comment]
One would not not use cdist or pdist for brute force search on a large data set. Write a small Cython (or C, C++, Fortran) function that does this with zero overhead. Also consider to multithread the function, e.g. using OpenMP.

**Comment by rgommers** [comment]
I agree that this is a large undertaking, and also that (given all the linked issues that @dkbarn posted two comments up) there is a clear demand for this. Unfortunately dtype support across scipy functionality tends to be pretty ad-hoc.

I'd say we in principle would like `float64` by default and also `float32` support everywhere where that's reasonable; we should avoid `float16` support as it's pretty niche.

**Comment by sturlamolden** [comment]
If we first start to template the dtype in C++ to support float32 then float16 is a very low-hanging fruit. Integer dtypes would be low hanging fruits as well. What we have to do is basically to replace ```double``` with a ```template<typename dtype>``` in the C++ code.

**Comment by rgommers** [comment]
Yes, I agree it'd be low-hanging fruit to add `float16` et al., but it's not a net win because of binary size. E.g.:
```
 ls -l build/scipy/spatial/_ckdtree.*.so
-rwxr-xr-x 1 rgommers rgommers 4930416 16 mei 08:25 build/scipy/spatial/_ckdtree.cpython-310-x86_64-linux-gnu.so
$ strip build/scipy/spatial/_ckdtree.cpython-310-x86_64-linux-gnu.so
$ ls -l build/scipy/spatial/_ckdtree.*.so
-rwxr-xr-x 1 rgommers rgommers 644880 18 mei 17:44 build/scipy/spatial/_ckdtree.cpython-310-x86_64-linux-gnu.so
```
So that's about 650 kb for `float64` only in release build mode. It may be justified to double that for `float32`, but I'd like to draw the line there, because binary package size is important too.
